File: src/main/java/alfheim/common/entity/spell/EntitySpellAcidMyst.java
package alfheim.common.entity.spell;

import alexsocol.asjlib.ASJUtilities;
import alexsocol.asjlib.math.Vector3;
import alfheim.AlfheimCore;
import alfheim.api.spell.*;
import alfheim.client.render.world.SpellEffectHandlerClient.Spells;
import alfheim.common.core.handler.CardinalSystem.PartySystem;
import alfheim.common.core.handler.SpellEffectHandler;
import alfheim.common.core.util.DamageSourceSpell;
import net.minecraft.entity.*;
import net.minecraft.entity.player.EntityPlayer;
import net.minecraft.nbt.NBTTagCompound;
import net.minecraft.util.AxisAlignedBB;
import net.minecraft.world.World;

import java.util.*;

public class EntitySpellAcidMyst extends Entity implements ITimeStopSpecific {
	
	public EntityLivingBase caster;
	
	public EntitySpellAcidMyst(World world) {
		super(world);
		setSize(1, 1);
	}
	
	public EntitySpellAcidMyst(World world, EntityLivingBase caster) {
		this(world);
		this.caster = caster;
		setPosition(caster.posX, caster.posY, caster.posZ);
		SpellEffectHandler.sendPacket(Spells.ACID, this);
	}
	
	@Override
	public void onEntityUpdate() {
		if (!AlfheimCore.enableMMO || (caster == null || caster.isDead || ticksExisted > 50)) {
			setDead();
			return;
		}
		if (this.isDead || !ASJUtilities.isServer()) return;
		
		List<EntityLivingBase> l = worldObj.getEntitiesWithinAABB(EntityLivingBase.class, AxisAlignedBB.getBoundingBox(posX, posY, posZ, posX, posY, posZ).expand(4.5, 4.5, 4.5));
		l.remove(caster);
		for (EntityLivingBase e : l) if (!PartySystem.mobsSameParty(caster, e) && Vector3.entityDistance(caster, e) < 9) e.attackEntityFrom(DamageSourceSpell.poison, SpellBase.over(caster, 1)); 
	}
	
	public int getTopBlock(World world, int x, int z) {
		int y = 255;
		while (y > 0 && world.isAirBlock(x, y, z)) --y;
		return y;
	}
	
	public boolean isImmune() {
		return false;
	}
	
	public boolean affectedBy(UUID uuid) {
		return !caster.getUniqueID().equals(uuid);
	}
	
	public void entityInit() {}
	
	public void readEntityFromNBT(NBTTagCompound nbt) {
		if (nbt.hasKey("castername")) caster = worldObj.getPlayerEntityByName(nbt.getString("castername")); else setDead();
		if (caster == null) setDead();
	}
	
	public void writeEntityToNBT(NBTTagCompound nbt) {
		if (caster instanceof EntityPlayer) nbt.setString("castername", caster.getCommandSenderName());
	}
}

File: src/main/java/alfheim/common/entity/spell/EntitySpellAcidMyst.kt
package alfheim.common.entity.spell

import alexsocol.asjlib.*
import alexsocol.asjlib.math.Vector3
import alfheim.AlfheimCore
import alfheim.api.spell.*
import alfheim.client.render.world.VisualEffectHandlerClient.VisualEffects
import alfheim.common.core.handler.*
import alfheim.common.core.handler.CardinalSystem.PartySystem
import alfheim.common.core.util.DamageSourceSpell
import alfheim.common.security.InteractionSecurity
import alfheim.common.spell.water.SpellAcidMyst
import net.minecraft.entity.*
import net.minecraft.entity.player.EntityPlayer
import net.minecraft.nbt.NBTTagCompound
import net.minecraft.world.World
import java.util.*

class EntitySpellAcidMyst(world: World): Entity(world), ITimeStopSpecific {
	
	var caster: EntityLivingBase? = null
	
	override val isImmune: Boolean
		get() = false
	
	init {
		setSize(1f, 1f)
	}
	
	constructor(world: World, caster: EntityLivingBase): this(world) {
		this.caster = caster
		setPosition(caster.posX, caster.posY, caster.posZ)
		VisualEffectHandler.sendPacket(VisualEffects.ACID, this)
	}
	
	override fun onEntityUpdate() {
		if (!AlfheimConfigHandler.enableMMO || caster == null || caster!!.isDead || ticksExisted > SpellAcidMyst.duration) {
			setDead()
			return
		}
		if (isDead || !ASJUtilities.isServer) return
		
		if (ticksExisted % 20 == 0) VisualEffectHandler.sendPacket(VisualEffects.ACID, this)
		
		val l = worldObj.getEntitiesWithinAABB(EntityLivingBase::class.java, alexsocol.asjlib.getBoundingBox(posX, posY, posZ).expand(SpellAcidMyst.radius)) as MutableList<EntityLivingBase>
		l.remove(caster!!)
		for (e in l)
			if (!PartySystem.mobsSameParty(caster!!, e) && Vector3.entityDistance(caster!!, e) <= SpellAcidMyst.radius && InteractionSecurity.canHurtEntity(caster ?: continue, e))
				e.attackEntityFrom(DamageSourceSpell.poison, SpellBase.over(caster, SpellAcidMyst.damage.D))
	}
	
	override fun affectedBy(uuid: UUID): Boolean {
		return caster!!.uniqueID != uuid
	}
	
	public override fun entityInit() {}
	
	public override fun readEntityFromNBT(nbt: NBTTagCompound) {
		if (nbt.hasKey("castername")) caster = worldObj.getPlayerEntityByName(nbt.getString("castername")) else setDead()
		if (caster == null) setDead()
	}
	
	public override fun writeEntityToNBT(nbt: NBTTagCompound) {
		if (caster is EntityPlayer) nbt.setString("castername", caster!!.commandSenderName)
	}
}